### PR TITLE
Libvirt Plugin

### PR DESCRIPTION
**Pull request by neckhair**
This Plugin collects virtdomain infos from a Libvirt deamon.
I'm planning to add interface and block device statistics in a later pull request.

```sh
$ ./telegraf -config etc/config.toml -test
* Plugin: libvirt, Collection 1
> libvirt_cpu_time,host=test,node=odin value=1447157114730809000i 1447157114730815923
> libvirt_max_mem,host=test,node=odin value=8388608i 1447157114730880494
> libvirt_memory,host=test,node=odin value=2097152i 1447157114730906774
> libvirt_nr_virt_cpu,host=test,node=odin value=2i 1447157114730921874
```
(Disclaimer: This is my first Go code. Feedback is much appreciated.)

### CODE REVIEW

**Comment by neckhair** [comment]
Any idea why the libvirt-go package can't be found by CircleCI?

**Comment by sparrc** [comment]
@neckhair you need to add it to the godep dependencies, like this:

```
go get github.com/tools/godep
godep restore ./...
godep save ./...
```

If you are having trouble using `godep`, you can give me write-access to your branch and I can fix it, or pull your changes and make a separate PR

**Comment by sparrc** [comment]
@neckhair thank you very much for this! looks good, could you please also update the supported plugins list on the README and add this to the CHANGELOG (under features)

File: plugins/libvirt/libvirt.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
+package libvirt
+
+import (
+	lv "github.com/alexzorin/libvirt-go"
+	"github.com/influxdb/telegraf/plugins"
+)
+
+const sampleConfig = `
+	# specify a libvirt connection uri
```
Comment (sparrc) [inline plugins/libvirt/libvirt.go]: these should be two-space indented

File: plugins/libvirt/libvirt_test.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package libvirt
+
+import (
+	"testing"
+
+	"github.com/influxdb/telegraf/testutil"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLibvirt(t *testing.T) {
```
Comment (sparrc) [inline plugins/libvirt/libvirt_test.go]: Does this test have dependencies? Does the host running the tests need to have libvirt installed?

File: plugins/libvirt/libvirt.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	for _, domainId := range domains {
+		domain, err := connection.LookupDomainById(domainId)
+		if err != nil {
+			return err
+		}
+
+		domainName, _ := domain.GetName()
+		tags := map[string]string{"host": domainName, "node": nodeName}
```
Comment (sparrc) [inline plugins/libvirt/libvirt.go]: Here you are overriding Telegraf's default `host` tag, is that intentional?

**Comment by neckhair** [comment]
@sparrc Thanks for your comments. I'll try to work on them in the next few days.

File: plugins/libvirt/libvirt.go (hunk 3 above)
Comment (neckhair) [inline plugins/libvirt/libvirt.go]: Not intentional, no. I'll rename it to "domain" which follows more the libvirt terminology.

File: plugins/libvirt/libvirt_test.go (hunk 2 above)
Comment (neckhair) [inline plugins/libvirt/libvirt_test.go]: The actual test does not need a running libvirt. But the compilation of go-libvirt needs the libvirt headers (libvirt-dev package). I added that package to `circle.yml`. Should I document that dependency anywhere else? 
For OS X there is a Homebrew formula for libvirt (`brew install libvirt`) which I have installed. That's why I didn't notice that fact earlier.

**Comment by neckhair** [comment]
@sparrc Ready for another review. Do you like all the commits to be squashed?

**Comment by sparrc** [comment]
@neckhair don't worry about squashing the commits.

I definitely know that some users would like this plugin, but I think I need to figure something out that won't make libvirt a build requirement. The reason being is that `libvirt` would be the first external (non-go) dependency added to the Telegraf project. Let me think and talk with some other Influx folks about what our options may be.

cc @otoolep @pauldix @toddboom thoughts on adding `libvirt` as an external dependency to Telegraf? Other build options we may have for "optional" plugins?

**Comment by otoolep** [comment]
Not a huge fan of adding an external dependency like this. Will I forced to get this library to build Telegraf? What if it doesn't exist for my platform?

**Comment by pauldix** [comment]
As long as we can still build for all platforms (Windows and ARM included) then I don't think it's an issue. Might make the builds more complex?

**Comment by sparrc** [comment]
@otoolep yes, you would be forced to, I'm trying to think of a way that a plugin could be optionally compiled

**Comment by sparrc** [comment]
@pauldix it means that building doesn't just require a Go environment, but also the `libvirt` C library

**Comment by neckhair** [comment]
Any news on this? If this gets merged I will work on network and disk stats.

**Comment by sparrc** [comment]
@neckhair I don't really like making libvirt a compile-time dependency of Telegraf, mainly because it means that we wouldnt be able to only use standard Go tools to build the project anymore, and also means that there would need to be an OS switch in the makefile for Ubuntu, Darwin, FreeBSD, CentOS, etc.

**Comment by neckhair** [comment]
Ok, I understand the concerns. I close this PR and we'll keep our own fork for internal use and reference for others. I created a new PR against our fork to keep track of it: https://github.com/ninech/telegraf/pull/3.

**Comment by sparrc** [comment]
@neckhair please re-open, I haven't gotten to updating the build process yet but still would like to include this in Telegraf

**Comment by neckhair** [comment]
@sparrc I'd like to but github doesn't let me (because I force-pushed?). Let me know when you have a way to do it. I'll create a new PR then.

**Comment by sparrc** [comment]
force pushes should be OK, I'll see what I can do

**Comment by sparrc** [comment]
@neckhair Do you think you can re-open this PR based off master?

It's probably easiest to just re-copy the plugin into the new `plugins/inputs` directory. I've got a `sensors` plugin that relies on `libsensor` now, so we can go forward with having plugins with C dependencies.

**Comment by Millnert** [comment]
@neckhair ( cc: @sofixa ): are you able to re-submit this as comments above mentions?
there are several users who would like to see this merged, including @sparrc. :)

**Comment by sofixa** [comment]
I'll give it a try later today.

**Comment by sofixa** [comment]
@sparrc I am currently working on rebasing the libvirt plugin with the current version of Telegraf, but i've hit something on the road:
The libvirt-go package @neckhair used no longer seems to be supported by the same person(the new repo is here https://github.com/rgbkrk/libvirt-go), and there are incompatibility issues with various versions of libvirt, so there are two packages  - v1 and v2(as tagged in GitHub) for versions 0.9.8 and below, and 1.2.2 and up, respectively (at least that's what i understand). 

How should i proceed? Import both versions and, depending on the libvirt version, use the one that corresponds? Or use only the v2, thus making the input unusable with older versions of libvirt?

**Comment by neckhair** [comment]
@sofixa @sparrc Thank you for working on this. Unfortunately I am busy with other stuff at the moment and can't help.

**Comment by gtt116** [comment]
@sofixa ASAIK libvirt-go v1 and v2 share the same API to get libvirt version, so we can using this api to detect which libvirt-go version I should used.  Also we should include both v1 and v2 into dependencies. Does it make sense?

**Pull request closed without merging**